lib/msgraph/client.go

```diff
@@ -0,0 +1,334 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package msgraph
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/defaults"
+)
+
+// baseURL is the default value for [client.baseURL]. It is the address of MS Graph API v1.0.
+const baseURL = "https://graph.microsoft.com/v1.0"
+
+// defaultPageSize is the page size used when [Config.PageSize] is not specified.
+const defaultPageSize = 500
+
+// scopes defines OAuth scopes the client authenticates for.
+var scopes = []string{"https://graph.microsoft.com/.default"}
+
+type azureTokenProvider interface {
+	GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error)
+}
+
+// Config defines configuration options for [client].
+type Config struct {
+	// TokenProvider provides tokens to authorize to MS Graph API.
+	// Concrete implementations of this are defined by [github.com/Azure/azure-sdk-for-go/sdk/azidentity].
+	TokenProvider azureTokenProvider
```

Comment:
```suggestion
	TokenProvider AzureTokenProvider
```

I would export the interface.

An exported field with an unexported type is awkward to use.